Adding eassert, eset, epose, etc.

herbelin opened the pull request:
This is a natural extension of the language of primitive tactics. It provides e-variants for `assert`, `pose`, `set`, `remember`, `pose proof`, `enough`, allowing to manipulate partial terms or statements, and hence to ease a way to develop proofs by refinement and unification. One can for instance see it as adding some benefits of "refine" to usual tactics (typically, I found the need for `eassert` when implementing `functional extensionality in`).

Not a commitment to the e-policy. With a translating tool, the e-prefix could eventually become the default if considered to be the good way to go. In the meantime, I however don't see other alternatives than adding specific e-variants. (See also #3872.)

Not a commitment to say that statements with free evars should be kept with unresolved evars rather than generalized over unresolved evars. For instance, Matthieu's `` ` `` could be generalized to tell "generalize" rather than "keep evars/fail", as in, say, providing ``assert `(_ = _)`` to be possibly interpreted as `assert (forall A (x y : A), x = y)`.

Not a commitment to say that the syntax of tactics should be hard-wired in `g_tactic.ml4`. I followed the easy path but maybe these tactics could already be definable with `TACTIC EXTEND` instead.

No documentation yet.

jonleivent commented:
@herbelin, you appear to have commitment issues ;)

Otherwise, cool!

As for generalizing vs. extentializing - I had a discussion with Pierre Courtieu IIRC about a tactic that was like eapply, but generalized over remaining args instead of making them holes.  Point being - it would be useful to have a general method that takes something that generates holes and instead turns those into generalizations.  An "unrefine"?

JasonGross commented on ltac/tacinterp.ml:
```diff
@@ -1745,37 +1747,37 @@ and interp_atomic ist tac : unit Proofview.tactic =
-            (TacLetTac(na,c,clp,b,eqpat))
-	    (Tacticals.New.tclWITHHOLES false (*in hope of a future "eset/epose"*)
+            (TacLetTac(ev,na,c,clp,b,eqpat))
+	    (Tacticals.New.tclWITHHOLES ev (*in hope of a future "eset/epose"*)
```
Shouldn't this comment be removed?

ejgallego commented:
This is likely a naïve question, but why are two sets of tactics needed? Why can't the the existing set, pose, etc... gain the capability of generalizing over evars?

JasonGross commented:
@ejgallego Backwards compatibility.  If you make `set`, `pose`, etc more powerful in this way, automation that previously failed will now succeed, and automation that previously fully solved a goal may no longer do so, because it would take a different path.

ejgallego commented:
@JasonGross umm, it looks to me that if automation would fail because of that, it would be very fragile automation to begin with.

I'm pretty convinced that supporting fragile automation is a futile goal.

maximedenes commented:
@herbelin do you want this PR to be discussed for integration on the 8.7 roadmap?

herbelin commented:
@maximedenes: yes, I think it is natural to have them, even if it gets eventually supersed by a new more flexible model. (And I miss them from time to time.)

Zimmi48 commented:
@herbelin It would be good if no new tactics were introduced without proper documentation. Not saying that this PR cannot get merged right now, but if it is then it should be on your TODO list to document the variants before the 8.7 beta release.

Another remark is that if we actually get an attribute system one day, then the e variants of tactics could instead be replaced by a uniform attribute, which could be set to emulate the behavior of the basic variant, of the e variant, or of the generalizing variant which was discussed above. And it could also be set more globally via an option...
But the fact that such attribute system is on the roadmap does not mean that we will effectively get it for 8.7 (right now no one is really working on it if I understood correctly; and there is not even a CEP), and thus, in the meantime, I support your approach.

herbelin commented:
@Zimmi48: doc: I agree. To be eventually superseded by a more uniform attribute system, no objection at all, on the contrary, since it would just become a particular case of a more general mechanism.

herbelin commented:
Fixed bugs and added doc. I remained vague on the specification of what a generated _existential variable_ is, but I think it is a general issue of the documentation, if not of the system in general, to have a clearer semantics on where (dependent) existential variables go (on the shelf? at which position? etc.).

ejgallego commented:
Hugo the PR is failing to build on Travis, could fix it?

herbelin commented:
@ejgallego: fix applied, now ok on travis.

Otherwise, on coqdev, @maximedenes asked about "What is the *long term* plan on e/not e-tactics? How is this a first step towards this long term plan?".

I answered that if the discussion is ripe for a long term plan on e-tactics I'm ok to be the one who organizes it. In particular, we can consider having an e-variant for each existing tactic, we can  consider providing a modifier to stop a tactic failing on unresolved evars, or consider making stopping failing on unresolved evars to be the default (did I miss other alternatives)?

But whatever the long term plan is, a first step is to allow each individual existing tactic to be able to behave in the e-way (in the sense of allowing unresolved evars).

So, whatever long term plan we have, I don't see how to avoid having this PR as a step towards it.

Also, I'm ready to extend the implementation of e-variants to other tactics, so that these are accessible using a generic modifier or general option, or whatever, but I'm not necessarily enthousiastic to add a concrete syntax for all of them.

More precisely: I think it makes sense, by regularity, to have `einversion`, `efix with`, `ecofix with`, `etransitivity c`, `eabsurd`, `egeneralize`, `estepl`, `estepr`, `especialize` but I would be more encline to deprecate the following than adding a `e`-syntax: `cut` (not only it is wrongly named, but most of the time it is used in a pattern of the form `cut;[intro|]` which suggests that `assert` is to preferred), `casetype` and `elimtype` (which in practise are used in a context of the form `casetype False`, or `elimtype False` which means that `exfalso` is the real intent), `cutrewrite` (already deprecated in favor of `replace`). Note that more thinking is needed about the expected behavior of `edestruct`/`einduction`, and that this is not unrelated to the `invert`-like tactic planned in ssr 2.0.

herbelin commented:
@maximedenes: did you take time to study my answer? To improve the fluidity of development, I really think that we should make progress on the PRs pending for a long time (congratulations in passing for today's efficiency!), and, unless I missed something in the discussion, this one is blocking on your answer.

ejgallego commented:
No strong opinion on this one, however as a user it doesn't feel great to have so many versions of tactics. In particular I was discussion with @Zimmi48 today how ssr `have` is versatile enough that replaces 4 (or in the case of this patch 8) tactics.

Zimmi48 commented:
There was some related discussion in https://github.com/coq/coq/pull/449#issuecomment-292924551 (where I remarked in particular that `pose proof` and `specialize` can basically be used to do exactly the same thing, however people seemed reluctant to merge these two) and some threads on coqdev as well.

@ejgallego `have` replaces `assert`, `pose proof` and `specialize` but I'm missing the fourth one.
Incidentally, all of these tactics are listed as "variants" of `assert` in the refman. See https://coq.inria.fr/refman/Reference-Manual010.html#hevea_tactic52. From what I can see in the code `pose proof` is a true variant of `assert` in terms of implementation, while `specialize` is distinct.

Now, I'd be interested to hear how this also replaces the e-variants proposed by Hugo here.
I tried using `have H : _ = _.` but this does not compare at all to `eassert (H : _ = _).` The first one generalizes each of the holes as a distinct variable while the second creates (as expected) one existential variable for each of the holes.

herbelin commented:
@ejgallego: It is certain that we can reduce the number of tactics, but
- Tactics that do not exists with the e-option do not support the e-semantics, so, implementing the e-semantics, even without giving it a syntax is a necessary preliminary step to eventually reduce the number of tactics, isn't it?
- We have to do with tactics as they are. We cannot change the semantics of non e-variants without  breaking the compatibility. Even if versioning tactics allows to support more semantics change of tactics, we should limit them so that users willing to upgrade can do it, shouldn't we?
- In my opinion, having tactic `foo` and tactic `efoo` has the drawback of blocking two entries in the parser but it is not so bad in terms of regularity and style compared to having `foo` and, say, a prefix`with_evar` to modify in the e-way the semantics of `foo`.

But, maybe are you actually talking about the number of tactics in general, independently of the existence of an e-variant?

In particular, when you say that ssr is versatile enough to replace 4 tactics, I guess that you are referring to `rewrite`? If so, wouldn't it be  fairer to say that ssr has symbolic notations for 4 tactics which all use the same `rewrite` keyword? My experience with symbolic notations is that it allows compactness and easy chaining but that it does not make a tactic call necessarily clearer, especially for a beginner, so even in the presence of symbolic notations, I think it makes sense to keep appealing clearly identified `keywords` that any reader, even if non-Coq user, can have an intuition about.

Also, we tried several times in the past to support in vanilla `rewrite` the possibilities of symbolic chaining that ssr `rewrite` provides but at the end, we were never really satisfied with the resulting syntax (`-` vs `<-`, the absence of commas in ssr's notation, etc.). So, it seems that to merge ssr `rewrite` and vanilla `rewrite`, some compromise at the syntax level is needed.

But if you think that such a merge of ssr `rewrite` within vanilla `rewrite` is possible, that would be interesting.

Zimmi48 commented:
@herbelin You misread a bit of @ejgallego comment. He's talking about the ssr tactic `have`.

maximedenes commented:
I think adding new tactics has some risk because of the number of already existing tactics, but is an ok compromise as a step towards a model that factors them using attributes for instance.

@herbelin Can you rebase this PR so that we merge it? Thanks!

herbelin commented:
@Zimmi48: Ah, sorry, I indeed misread @ejgallego's comment.

It happens that vanilla `assert` do have the syntax `assert (H:=foo)` which does `pose proof`!

Actually, it is even the other way round, `pose proof` has been introduced because the syntax `assert (H:=foo)` was considered to be too ad hoc and a bit unappropriate.

So, maybe, ssr, which roughly derived from vanilla Coq as it was at the beginning of the 2000's, just kept vanilla's `assert` as it was at this time ("pose proof" is from 2006).

herbelin commented:
@Zimmi48, @ejgallego: By the way, it is a long time that I wonder about the user feedback wrt the 
"generalization over holes as distinct variables" feature. It looks interesting and I would be happy to know more about the use cases.

Zimmi48 commented:
@herbelin From what Georges Gonthier told me yesterday, `have` was introduced when Coq had only `cut`, thus before vanilla `assert`.

herbelin commented:
@Zimmi48: The tactic `assert` was introduced in 2001. I suspect that the name `have` came after `assert` was implemented since my memories is that this was precisely one of the reasons of friction between Georges and the Coq developers at this time, that he renamed `assert` into `have` without a strong motivation (I vaguely remember that he wanted to get rid of the Coq's constraint that terms arguments of tactics are always surrounded by parentheses; maybe he also found `have` more appealing that `assert`, I don't know; the motivation for the divergence is still unclear to me as of today.)

Maybe it is time that Georges and the Coq developers from the 2000's meet and get beyond the syntactic quarrel from this period of time, concentrating on how to work together rather than maintaining a futile "schism", as P. C. called it.

herbelin commented:
@Zimmi48: We have `Definition/Theorem`, `Variable/Hypothesis`. This is redundant from the CIC point of view, but we keep it: one puts the focus on the term, and the other on the type; they also corresponds to different intuitions. Similarly, all of `pose`, `set`, `assert`, `pose proof`, and to some extend `enough`, are redundant, they are all a `let` from the CIC point of view. The choice of names between `pose` and `set` is a bit unfortunate (reverting them would better match the usage I think). But all of them are justifiable in that they highlight a different use of the `let`. Two are proof-irrelevant out of which one has a focus on the type and the other on the proof term.

This being said, what you propose makes sense. I'm ok to undeprecate `assert (H:=proof)`.

For `specialize`, I believe the following does what you want:
```coq
Lemma L (H:True->True) : True.
assert (H:=H I).    
```

Zimmi48 commented:
@herbelin I'm not arguing against having synonyms. I like synonyms. But I like symmetry as well. Two given variants should behave similarly in every point but the point why they are different.
But OK, let's talk again about this proposal.

herbelin commented:
@maximedenes: Rebased and documentation added. We merge it?

maximedenes commented:
I'll let the test run one last time and merge it.

ejgallego commented:
@Zimmi48 the 4th I was thinking was `cut`, but I am not sure.

Regarding `have`, my impression is that George was talking of the first macro language used in the proof  of the 4 color theorem, they had an `HAVE` macro I believe, and that would predate `assert` in Coq.

Unfortunately I cannot find that technical report now describing that macro language now and I have to go. My understanding is that later on, changes in Coq made the implementation of the macro language impossible and thus the plugin was born.

By the way I am puzzled by:
> eassert (H : _ = _).

making `_` to be the same variable. That seems a bit strange doesn't it? How can I assert the typical use case with `have`, `have := IHn _ _ h1 _ h2` ?

Zimmi48 commented:
@ejgallego No indeed, this generates two distinct existential variables (I was saying it generates one for each hole).

ejgallego commented:
Oh @Zimmi48 I misunderstood, sorry for the noise I guess I am not having a good day today so I can't even properly read english. By the way my original remark was not meant against this PR, but more in general about the roadmap. For instance, I was thinking if instead of adding these new tactics modeled after their old counterparts in this PR, we would add a single, general and more powerful forward reasoning tactic. In other words, do you want users to migrate to the new "etactics", that may become deprecated later, or to aim for a more general forward reasoning tactic ?

@herbelin I think there are many uses of the generalization features in `have` and variants. I like it as:
`have H := IH _ p` won't for instance create an evar with the scoping problems this brings sometimes.

Zimmi48 commented:
Thinking more about this PR, I agree that:
1. `eassert` is useful to start proving sub-goals which we do not want to specify completely.
2. `eset` is useful because `set` does not work on terms containing existential variables currently.

However, I'm much more doubtful about `epose proof`. In the case of this tactic (and similarly for `specialize`), I don't see why when we want to specify the second argument (or hypothesis), the first one should become an existential variable. In that case, the generalization behavior of the `have` tactic seems more natural (since we are not losing anything by generalizing over holes when we specialize a term). We can still use `eapply` later to replace unspecified arguments by existential variables.

However, I would argue against calling `epose proof` (or `especialize`) a tactic which does this generalization (and has nothing to do with evars). I would instead just make the current `pose proof` / `specialize` / `assert ( := )` tactics more powerful (and I don't think this would break anything).

Furthermore, the current implementation gives strange names to the new existential variables, see on this example:
```
Goal (forall n m, n <= m -> m <= n -> n = m)
  -> True.
intro H.
epose proof (H _ 3) as H.
```
gives:
```
1 subgoal
H : ?Goal <= 3 -> 3 <= ?Goal -> ?Goal = 3
______________________________________(1/1)
True
```

As a first step, you could just remove the `epose proof` tactic from this PR (unless you have a use case in mind of course). What do you think?

PS: I did not test `epose`, `eremember` and `eenough` but it seems to me that `eenough` is useful for the same reasons `eassert` is useful, `eremember` is probably useful for the same reason `eset` is useful. And I don't recall what `pose` is about.

JasonGross commented:
I sometimes instantiate arguments to hypotheses with evars so that I can (a) see what the next argument is from ltac or (b) rewrite in one of the side conditions that was previously under a binder without having to invoke `setoid_rewrite` (which is often slow).

herbelin commented:
(For the sake of completeness of the discussion, even if probably everyone interested would have checked by him/herself, ssr's `have:` also follows the policy of using the goal as working place, so `have: T` is `cut T; swap 1 2` and `suffices: T` is `cut T`. In practice, most uses of `cut` are of the form `cut T; [intro|]`, from what I draw the conclusion that `cut` is maybe not a so interesting brick.)

@Zimmi48: I'm not really convinced by the interest of removing `epose proof`, since the idea is precisely to go towards more uniformity, not to more singularity. On the other hand, even in the presence of `epose proof`, it would still be possible to extend `pose proof` so that it does a generalization. So `epose proof` does not remove anything to your idea.

An alternative to generalizing evars could be to generalize unresolved evars rather than free names, as we do in mathematics. If, to be explicit on the intent, we expect an explicit modifier, say @mattam82's `` ` ``, one could do that ``pose proof `(Nat.add_assoc x 2 x)``, for `x` not in the context, adds an hypothesis `forall x, x + (2 + x) = (x + 2) + x`. On the other side, this kind of feature (marginally) weakens the name robustness since it is context sensitive... It seems that generalizing evars may be more convenient for implicit polymorphism while generalizing variables seems more natural, in particular to express sharing.

About the name `?Goal`, we can certainly do better. I will look at it.

Zimmi48 commented:
@JasonGross How do you do that currently? It looks anyways like hacks which are used to avoid current limitations of Ltac / rewrite but not like they are giving a fundamental usefulness to the `epose` tactic.

@herbelin OK, so here's another proposal for making things more uniform. Instead of having these new tactics `epose proof` and `eassert`, let's just make the existing `pose proof` and `assert` more powerful by supporting `_` holes with the following rule: a `_` in a type means create an evar unless the hole can be inferred; a `_` in a term means generalize over this hole, unless the hole can be inferred. With this approach, `pose proof` and `assert ( := )` can be used for "generalized partial application".

As for `set`, `_` holes are already supported to create patterns to be matched against a sub-term of the current goal. This is not the same thing IMO. It is more related to patterns which can also be passed as argument to `simpl` or which might soon be used with `rewrite`. That these patterns do not match sub-terms containing evars, while they do in the case of `simpl` is unfortunate and should be fixed.

The only reason that I can see for not going this route would be compatibility with existing developments. But I feel like at least the first part of my proposition shouldn't break anything (simply because these tactics are currently not used with holes).
The second part could break some stuff (if a different pattern is selected because `set` became more powerful) but that's not very likely and it could be tested with Travis.

In any case, I don't feel already having `eapply`, `eassumption` and `eauto` to be a compelling argument to add more e-tactics. You said it yourself in the introducing comment:

> Not a commitment to the e-policy. With a translating tool, the e-prefix could eventually become the default if considered to be the good way to go. In the meantime, I however don't see other alternatives than adding specific e-variants. (See also #3872.)

To insist a bit more on the difference with `apply` vs `eapply`: the behavior of the two is different when applied in the same way. One may decide to generalize `apply` to support `_` holes. But then `apply` would typically be used in the following way: `apply with _.` or `apply with (x := _).` So IMHO `eapply` is here to stay.
One could similarly make `exists _.` be synonymous to `eexists.`
As for `eassumption`, the main reason for keeping it distinct from `assumption` is that it can take decisions which would then affect the provability of other sub-goals (it does not create new evars, it solves them).
And finally I'm more doubtful on the usefulness of `auto` vs `eauto` (apart from the very obvious compatibility problems which would be created by making the former a synonymous of the latter).

Note that this proposal is very close to what @RalfJung was proposing somewhere (can't find the reference anymore).

JasonGross commented:
@Zimmi48 I don't recall where to look in my developments for an example in the wild, but here's a kind of dumb tactic that does the kind of thing I'm talking about (with `specialize` rather than `pose proof`):
```coq
Ltac especialize_nat :=
  repeat match goal with
         | [ H : forall n : nat, _ |- _ ] => let n' := fresh n in evar (n' : nat); specialize (H n')
         end.
```

Also, I expect the first part of your proposal to break some of my code (at least if you do it with `specialize` and not just `pose proof`).  I do things like
```coq
match goal with
| [ H : forall x y z w, _ = _ -> _, H' : _ = _ |- _ ]
  => specialize (H _ _ _ _ H')
...
end
```
and rely on this branch to fail in cases where the `x`, `y`, `z`, and `w` can't all be inferred from the type of `H'`

maximedenes commented:
Summary of WG discussion: we merge this (after trivial rebase)

herbelin commented:
@Zimmi48: you were not at the discussion. Are you ok as it is?

herbelin commented:
I rebased and added a refinement to address an early [comment](https://github.com/coq/coq/pull/248#issuecomment-298174512) of @Zimmi48 on cool evar naming (now displaying `?n` rather than `?Goal`). Maybe @ppedrot wants to have a look to tell if it is consistent with his view at evars.

The extra commit overlaps with the `Sigma` (#623), so this can wait for it.

Zimmi48 commented on doc/refman/RefMan-tac.tex:
```diff
@@ -1155,6 +1155,13 @@ \subsection{\tt set ( {\ident} := {\term} )}
+
+  While the different variants of \texttt{set} expect that no
+  existential variables are generated by the tactic, \texttt{eset}
+  removes this constraint.
```
Could you maybe provide an example when this could happen? I don't know if it would be worth including it in the doc but at least for my own understanding.

BTW, last time I tried this PR I discovered that `eset` was indeed more powerful than `set` because the latter does not work on terms containing existential variables but we agreed this qualified as a bug of `set`. Did you have time to look into it?

Zimmi48 commented on doc/refman/RefMan-tac.tex:
```diff
@@ -1170,6 +1177,15 @@ \subsection{\tt set ( {\ident} := {\term} )}
+  {\tt eremember {\term} as {\ident} eqn:{\ident}}
+
+  While the different variants of \texttt{set} expect that no
+  existential variables are generated by the tactic, \texttt{eset}
```
Failed copy-paste here.

Zimmi48 commented on doc/refman/RefMan-tac.tex:
```diff
@@ -1284,6 +1308,13 @@ \subsection{\tt assert ( {\ident} :\ {\form} )}
+
+  While the different variants of \texttt{assert} expect that no
+  existential variables are generated by the tactic, \texttt{eassert}
+  removes this constraint.
```
I would add: this allows not to specify the asserted statement completely before starting to prove it.
To me, this is the most useful feature provided by this PR. It should be described more positively.

Zimmi48 commented on doc/refman/RefMan-tac.tex:
```diff
@@ -1320,6 +1356,14 @@ \subsection{\tt assert ( {\ident} :\ {\form} )}
+
+  While the different variants of \texttt{eenough} expect that no
+  existential variables are generated by the tactic, \texttt{eenough}
+  removes this constraint.
```
Typically, `eenough (P _)` can be used to replace `enough (exists x, P x) as []; [| eexists ].`

maximedenes replied:
first occurrence of `eenough` should be `enough`.

Zimmi48 replied:
@maximedenes Judging by the rest of the documentation, this is a typo and Hugo meant to say `enough` instead of `eenough` in the first occurrence of the two.

maximedenes replied:
Yes, that's why I was mentioning it.

herbelin commented:
@maximedenes, @Zimmi48: Ah, thanks a lot. Fixed and now rebased with @ejgallego's move of loc outside of ast.

@Zimmi48: For `eset`, I feel that it is needed in anticipation of a general e-modifier to be interpreted for every tactic, but maybe should it fail then, telling to use `epose` instead? Told otherwise, in the presence of a general modifier `with_evar tac`, what should `with_evar set (x:=t)` do?

Zimmi48 commented:
My question was: is `eset` more powerful than `set`? From your answer, I seem to understand it isn't really more powerful (if the aforementioned bug of `set` is fixed). Does someone have an idea what `eset` could be used for that `set` cannot be?

If indeed `eset` and `set` have exactly the same expressiveness, then it is confusing to have the two. Of course, in the presence of a generic modifier, it would be expected that it has no effect on some tactics (e.g. `with_evar idtac` should behave the same as `idtac`).

But maybe the aforementioned bug is not a bug (I don't mind if we decide it isn't). In that case the `eset` tactic *does* have a purpose but its purpose should be reflected better in the doc.

For reference the bug in question is the following:

```
Goal exists n : nat, n = n -> True.
eexists.
set (H := _ = _).
(* fail with Error: Cannot infer this placeholder of type "nat". *)
```

herbelin commented:
@Zimmi48: I pushed an extra commit to fix the bug that you are mentioning. Somehow, one should try to push it to v8.6, to see what Travis says.

Continuing about `eset`, on Wednesday afternoon at the WG, there was a discussion on tactic options in relation with #420 (see also #677), and in particular about whether it would be better to have general tacticals (say `with evars in tac` for any `tac`, or `with structural injection in tac`) or to rather attach options/attributes to specific tactics (say `set (H := _ = _) [with evars]` or `set ~with_evars (H := _ = _)`, or `set (H := _ = _) (options: with evars)`, or `set (H := _ = _) (options: with evars) with option evars`, or whatever cool syntax we find). In the latter case, it would indeed make sense not to have `eset` at all (i.e. simply not to have the `with evars` option available for `set`).

ppedrot commented:
@herbelin @Zimmi48 From my very implementation-driven point of view, this whole e-or-not variant of a given set of tactics completely misses the point in most instances. Technically, the part of the semantics that changes is not in the tactic, but in the interpretation of the constr argument passed to the tactic. This is clear in the code, because what makes the difference is the use of the `tclWITHHOLES` tactical around the provided constr.

Thinking in an effectful ML, we would have the following definitions and signatures for e.g. the `set` tactic:
```ocaml
val prim_set : ident -> const -> unit
(** Primitive set *)
val with_holes : (unit -> 'a) -> 'a
(** Evaluate a thunk and check it does not generate undefined evars *)

let set id c = let c = with_holes c in set id c
let eset id c = set id (c ())
```
So really, the difference in this case is attached to the argument, not to the tactic.

maximedenes commented:
During the discussion at the WG, we said this PR was only adding tactics, not changing the behavior of existing ones. Is it still true after the `tclWITHHOLES` change?

Zimmi48 commented:
@maximedenes 
> During the discussion at the WG, we said this PR was only adding tactics, not changing the behavior of existing ones. Is it still true after the tclWITHHOLES change?

This isn't true anymore but this should be viewed as an independent bugfix that @herbelin could have pushed to a different PR (but is also suitable to have here to avoid cluttering more the PR system right now IMHO).

maximedenes commented:
> could have pushed to a different PR (but is also suitable to have here to avoid cluttering more the PR system right now IMHO).

As I said during the working group, separating independent commits and pushing them to separate PRs actually makes the whole thing faster, because it makes sure the changes that raise no issue (w.r.t. discussion and tests) are merged very quickly.

herbelin commented:
@Zimmi48: Sorry, I just saw now your comments on the doc. I made an update. Please tell if ok for you.

@maximedenes: The fix is indeed by definition a change of semantics. I'm waiting for Travis to know what to do with it. Eventually, it should rather go to v8.6 I guess and I'll do a proper PR for it.

herbelin commented:
@ppedrot: I don't know if it is so clear. For instance in `apply f, g, h`, the current semantics is that `f` and `g` are allowed to introduced evars if `h` eventually solved them (something which @silene actually found surprising but which I find rather convenient in practice). Similarly, in `destruct (S _)` or `set (S _)`, the constr is matched against a subterm of the goal, so evar-behavior is also attached to the tactics. Also, `eauto` produces orphan evars in some cases. We may want to forbid that even though there is no constr argument around. (Alternatively, we can of course also adopt the view that it is a `eauto` bug.)

Zimmi48 commented:
@herbelin I think you forgot to push your update to the doc.

Zimmi48 commented:
Thanks! This now looks great to me.

maximedenes commented:
@herbelin Travis passed. Could you isolate the commits that are really implementing what the title of this PR suggests? We could then merge it. Thanks!

herbelin commented:
@maximedenes: I made #693 for the fix.

If we want to nitpick, there is also "Support for using type information to infer more precise evar sources." which, in rare circumstances, changes the name the user sees for evars whose name is generated by Coq (typically evars associated to a dependent `_` are named after the name of the variable they are bound to rather than `?Goal`). I can make a different PR for it if someone wants to specifically comment on that.

maximedenes commented:
@herbelin Thanks!

As for "Support for using type information to infer more precise evar sources.", if everyone is ok with it (@ppedrot?), we can leave it here. What I'd like to avoid is that the entire PR gets delayed if a bit of discussion is needed on one item.

maximedenes commented:
Something is wrong, the test-suite does not pass.

ppedrot commented:
LGTM.

herbelin commented:
@maximedenes: Just a shift in evar ID for emacs. Fixed.

By the way, there is another problem with test-suite on MacOS X since #687: `readlink -f` does not work on MacOS X. Setting `BIN := $(shell cd ../bin; pwd)/` is a way e.g. to make it works. @SkySkimmer: would you be ok to do something for it?

SkySkimmer commented:
@herbelin see #701

herbelin commented:
@SkySkimmer: thanks.
